File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.py
class Solution(object):
    def sumByD(self, nums, div):
        sum=0
        for num in nums:
            sum += (num + div - 1)// div
        return sum
    def smallestDivisor(self, nums, threshold):
        """
        :type nums: List[int]
        :type threshold: int
        :rtype: int
        """
        if len(nums) > threshold: 
            return -1
        low=1
        high = max(nums)
        #BS on answer
        while(low< high):
            mid = low + (high-low)//2
            if self.sumByD( nums, mid)<=threshold :
                high = mid
            else:
                low = mid+1
        return low

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.js
/**
 * @param {number[]} nums
 * @param {number} threshold
 * @return {number}
 */

var sumByD= function(nums, div){
    let sum=0;
    for (let num of nums){
        sum += Math.floor((num + div - 1) / div);
    }
    return sum;
}
var smallestDivisor = function(nums, threshold) {
    if(nums.length > threshold){
        return -1;
    }
    let low =0;
    let high = Math.max(...nums);
    //BS on answer
    while (low < high){
        let mid = Math.floor((low + high) / 2);

        if (sumByD(nums, mid) <= threshold){
            high = mid;
        } else {
            low = mid + 1;
        }
    }return low;
};